### ISSUE TITLE
[DOC] Guide 'structure-widgets.md' lacking info for passing config variables

### ISSUE THREAD

**Issue by WadeShuler** [yiisoft/yii2#9818]
The docs guide on "structure-widgets" is aimed to teach us how to create a widget. It is missing one big thing though. Many widgets need default config options. While the example shows that we can pass them when using the widget in a view, it doesn't tell the end user how to pass default params in their config file.

A prime example of when/why this is necessary, is the [yii2-recaptcha-widget](https://github.com/yiisoft/yii2/blob/master/docs/guide/structure-widgets.md). We don't want to include our site key and secret every single time we use the widget.

I know how to do it :) I just think the docs should show people how.

Link:
[https://github.com/yiisoft/yii2/blob/master/docs/guide/structure-widgets.md](https://github.com/yiisoft/yii2/blob/master/docs/guide/structure-widgets.md)

**Comment by samdark** [comment]
It's currently described at https://github.com/yiisoft/yii2/blob/master/docs/guide/concept-di-container.md#practical-usage-.

**Comment by WadeShuler** [comment]
I think it should at least be referenced on the page I mentioned. If I was to try and find out how to make my own widget, that is the page I would have checked out for more info, and been let down.

**Comment by samdark** [comment]
I've added a note.

**Comment by WadeShuler** [comment]
That wasn't what I was talking about. I haven't seen that implementation... common would be adding to your config `main.php` under components array..

**Comment by samdark** [comment]
Umm... that's doesn't look like a common widget config. It should be in this particular widget docs.

**Comment by WadeShuler** [comment]
Go look at yii2 recaptcha widget. You set your google API credentials in you main.php and use the widget in your view, without having to redeclare them again.

**Comment by WadeShuler** [comment]
Example widget, based on the example on the page I linked to in the docs in my first post.'. This is exactly the same, only it takes the doc 1 step further.

In your configuration components array, add your widget and params. `class` maps to your class file, and `message` is a public param that we are configuring globally.

```
'components' => [
    'helloWidget' => [
            'class' => 'common\widgets\HelloWidget',
            'message' => 'My Config Message',
        ],
],
```

app/components/HelloWidget.php

```
<?php
namespace app\components;

use yii\base\Widget;
use yii\helpers\Html;
use Yii;

class HelloWidget extends Widget
{
    public $message;

    public function init()
    {
        parent::init();
        if ($this->message === null) {
            $this->message = Yii::$app->helloWidget->message;
        }
    }

    public function run()
    {
        return Html::encode($this->message);
    }
}
```

views/site/index.php

```<?php
use app\components\HelloWidget;
?>

<?= HelloWidget::widget() ?>
```

It passes `message` from the config to the widget.

I understand that in "other" places in the docs, it talks about configurations. However, this page is talking about creating a widget. This is the "Creating Widgets 101" page... It should also provide a simple example on how to set default configuration (ie: in main.php) to use throughout your app. Then a note with a link, to more information on configurations.

Leaving things like this out, is a common issue throughout the docs. It talks about something, but doesn't provide all the details for it, and doesn't provide a resource link to accomplish it. Causing the user to have to jump through hoops, and frustration with the docs.

Point being. If I was just getting started with Yii2, and I wanted to create my own widget, this page is where I would end up.

What I am illustrating here is extremely common. The yii2 recaptcha widget is a prime example. Other possibly applications, a Google Maps widget. I wouldn't want to place my Google API credentials in a million places, so setting it in my config main.php. Another example, Social Media share buttons widget. Set your credentials in your main.php components array, and your widget uses them globally.

The DI Container is a way to do this programmatically. The way I am talking about, is via placing the information (params) into your config directly. 99% of every extension I have used, I have had to add config details into my `main.php`.
